### ISSUE TITLE
Typing in strings become very slow with semantic-mode enabled

### ISSUE THREAD

**Issue by Yevgnen** [jorgenschaefer/elpy#1462]
# Summary

Thank you for reporting to Elpy !

Here is a template to help you do it so your issue can be fixed as
soon as possible. Please replace this paragraph with a summary of
your issue and fill the next sections.

# Steps to reproduce
Open a new Python file, typing something(or press a key for a while) in the strings.

Disabling either `elpy-mode` or `semantic-mode` solved the issue. I'm not sure whether it's `elpy`'s issue...

`profiler-start` with `cpu` gives 

```
- ...                                                            1397  76%
   Automatic GC                                                  1391  76%
 + minibuffer-complete                                              4   0%
 + elpy-rpc                                                         2   0%
+ timer-event-handler                                             253  13%
+ command-execute                                                 160   8%
+ redisplay_internal (C function)                                  11   0%
```

# My configuration
## OS
`macOS 10.13.4`

## Result of `(elpy-config)`
```
Virtualenv........: None
RPC Python........: 3.6.6 (/usr/local/var/pyenv/shims/python)
Interactive Python: python3 (/usr/local/var/pyenv/shims/python3)
Emacs.............: 26.1
Elpy..............: 1.24.0
Jedi..............: 0.12.1
Rope..............: 0.10.7
Autopep8..........: Not found (1.4 available)
Yapf..............: 0.22.0 (0.24.0 available)
Black.............: Not found (18.6b4 available)
Syntax checker....: flake8 (/usr/local/var/pyenv/shims/flake8)
```

## Elpy configuration in my init.el
```
(package-initialize)
(require 'use-package)
(require 'company)
(setq company-idle-delay 0.1
      company-echo-delay 1
      company-auto-complete t
      company-auto-complete-chars "?<>:\"'"
      company-require-match nil
      company-minimum-prefix-length 2
      company-show-numbers t
      company-selection-wrap-around nil
      company-tooltip-align-annotations t
      company-dabbrev-minimum-length 2
      company-dabbrev-code-modes t
      company-dabbrev-other-buffers t
      company-backends '(company-files
                         (company-keywords
                          company-capf
                          company-dabbrev-code)
                         company-dabbrev
                         company-yasnippet)
      company-transformers '(company-sort-by-backend-importance)
      company-global-modes '(not message-mode help-mode inferior-python-mode))
(global-company-mode)

(require 'semantic)
(add-hook 'python-mode-hook #'semantic-mode)

(use-package elpy
  :ensure t
  :diminish elpy-mode
  :after python
  :init
  (setq elpy-modules
        '(elpy-module-company
          elpy-module-eldoc
          elpy-module-pyvenv
          elpy-module-sane-defaults)
        elpy-shell-echo-input nil
        elpy-shell-echo-output 'when-shell-not-visible)

  :config
  (elpy-enable))
```

**Comment by galaunay** [comment]
Could be due to #1403.

I will make the feature optional, to avoid that kind of issues.

**Comment by Yevgnen** [comment]
@galaunay Thanks for for the quick response. I've updated `elpy` to `elpy-20180916.839`, the issue seems still there. Using `describe-variable`, I confirm that `elpy-company-add-completion-from-shell` is already set to `nil`. Maybe they are not related.

**Comment by galaunay** [comment]
Does it still happen with company mode disabled (or `company-idle delay` set to 1) ?

**Comment by Yevgnen** [comment]
The issue is gone if `company-mode` is disabled, while setting a large value (e.g. `1`) for `company-idle delay` does a bit help.

**Comment by galaunay** [comment]
Company-mode should be disabled when you disable Elpy.
Does the problem still exist if you deactivate Elpy but then re-activate company-mode ?

If it is the case, I guess the problem would come from a bad interaction between semantic and company.

If it is not the case, I guess it would be due to one of the company parameters modified by Elpy (at line 3395 in elpy.el):
```Elisp
     (set (make-local-variable 'company-idle-delay)
          0.01)
     ;; And annotations should be right-aligned.
     (set (make-local-variable 'company-tooltip-align-annotations)
          t)
     ;; Also, dabbrev in comments and strings is nice.
     (set (make-local-variable 'company-dabbrev-code-everywhere)
          t)
     ;; Add our own backend and remove a bunch of backends that
     ;; interfere in Python mode.
     (set (make-local-variable 'company-backends)
          (cons 'elpy-company-backend
                (delq 'company-semantic
                      (delq 'company-ropemacs
                            (delq 'company-capf
                                  (mapcar #'identity company-backends))))))
```

**Comment by Yevgnen** [comment]
`company-mode` is disabled if `elpy-mode` is disabled.

The problem does not exist if I deactivate Elpy but re-activate company-mode, the value of `company-backends` is 

```
(company-files
 (company-keywords company-capf company-dabbrev-code)
 company-dabbrev company-yasnippet)
```

Did you see the 'update' part in my original post? An IPython process buffer also causes this issue, so I'm not sure if it is related to semantic mode.

**Comment by galaunay** [comment]
Alright, so it seems to me that it is linked to the way Elpy modify company behavior.
So it should be gone if you deactivate `elpy-module-company`, i.e. using:
```Elisp
(setq elpy-modules
        '(elpy-module-eldoc
          elpy-module-pyvenv
          elpy-module-sane-defaults)
        elpy-shell-echo-input nil
        elpy-shell-echo-output 'when-shell-not-visible)
```
Is it happening ?

**Comment by Yevgnen** [comment]
It's gone...but I want completion....

**Comment by galaunay** [comment]
Yeah, sure.
I am just trying to figure out where the problem comes from.

As a temporary workaround, you should be able to get the completion working by deactivating the company module (last above) and activating company manually in your config file with:
```Elisp
(add-hook 'python-mode-hook
          (lambda ()
            (set (make-local-variable 'company-backends)
                 (cons 'elpy-company-backend
                       (mapcar #'identity company-backends)))
            (company-mode 1)))
```

**Comment by Yevgnen** [comment]
@galaunay Works like a charm~

**Comment by Yevgnen** [comment]
Hi, it seems `(set (make-local-variable 'company-idle-delay) 0.01)` caused the issue. If I remove it, everything just fine. My personal original setting is `0.1`.

**Comment by galaunay** [comment]
Thanks for te feedback.

After a quick try with a value of `company-idle-delay` of `0.1`, it doesn't seems to be really slower. 
Maybe this would be a better default value that won't slow down Emacs.

Did you run into this issue on a particularly not powerful computer ?

**Comment by Yevgnen** [comment]
😅I'm on a MacBook Pro 2014 running with macOS 10.13... Yesterday the issue was still here (if `0.01`) after a completely fresh installation of macOS 10.14...

**Comment by galaunay** [comment]
It should indeed not happen...
I don't know if it is not hiding another issue, it should work fine with `0.01`...

I will try to have a look.

In the meantime (that can be quite a long time...), you can reactivate `elpy-company-module` and overwrite elpy customisation by adding `(add-hook 'elpy-mode-hook (lambda () (setq company-idle-delay 0.1)))` in your config file.

**Comment by galaunay** [comment]
I have nothing against using 0.1 as the default value.
I use this value myself and I think it is already fast enough.

@jorgenschaefer, as the person who put 0.01 in the first place, what do you think ?
Additional question: do you think the problem may be that the completion mechanism has became slower than before ?

**Comment by jorgenschaefer** [comment]
I really like the "suggestion as you type", and felt 0.1 was a bit slow for me, but I'm fine either way … And not sure if this got slower, but maybe some kind of debouncing or cancelling old callbacks would already work maybe? Ideally, even multiple callbacks should not make Emacs noticably slower, I don't even know why anyone would notice the callbacks queuing up.

**Comment by galaunay** [comment]
I put `company-idle-delay` to `0.1` by default (PR #1526), as the "default value that would roughly work for everyone" .

I also opened an issue (#1527) to investigate those slow downs and eventually go back to faster suggestions.

I need to check if it can be due to my latest additions to the completion system.

**Comment by isolet** [comment]
@mpereira I agree that user custom value `company-idle-delay` should not be overwritten generally. For user who set it to `1.0`, it seems fast as the comment in the code says ` ;; We want immediate completions from company.`. However, for user who set it to `0.02`, it gets slow down if it is set again to `0.1` and need to set back to `0.02`.
